Algoritmo Random Forest

In [233]:
# Importação das bibliotecas
import pandas as pd
# Dividir os dados entre treino e teste
from sklearn.model_selection import train_test_split
# Realizar o label encoder
from sklearn.preprocessing import LabelEncoder
# Criar a matriz de confusão e obter a acurácia
from sklearn.metrics import confusion_matrix, accuracy_score
# Algoritmo Random Forest
from sklearn.ensemble import RandomForestClassifier

In [234]:
# Leitura da base de dados
database = pd.read_csv('credit.csv')
# Exibe as primeiras cinco linhas da base de dados
database.head()

checking_status  duration                    credit_history  \
0              <0         6  'critical/other existing credit'   
1        0<=X<200        48                   'existing paid'   
2   'no checking'        12  'critical/other existing credit'   
3              <0        42                   'existing paid'   
4              <0        24              'delayed previously'   

               purpose  credit_amount      savings_status employment  \
0             radio/tv           1169  'no known savings'        >=7   
1             radio/tv           5951                <100     1<=X<4   
2            education           2096                <100     4<=X<7   
3  furniture/equipment           7882                <100     4<=X<7   
4            'new car'           4870                <100     1<=X<4   

   installment_commitment       personal_status other_parties  ...  \
0                       4         'male single'          none  ...   
1                       2  'female div/dep/mar'          none  ...   
2                       2         'male single'          none  ...   
3                       2         'male single'     guarantor  ...   
4                       3         'male single'          none  ...   

    property_magnitude age  other_payment_plans     housing existing_credits  \
0        'real estate'  67                 none         own                2   
1        'real estate'  22                 none         own                1   
2        'real estate'  49                 none         own                1   
3     'life insurance'  45                 none  'for free'                1   
4  'no known property'  53                 none  'for free'                2   

                    job num_dependents  own_telephone foreign_worker class  
0               skilled              1            yes            yes  good  
1               skilled              1           none            yes   bad  
2  'unskilled resident'              2           none            yes  good  
3               skilled              2           none            yes  good  
4               skilled              2           none            yes   bad  

[5 rows x 21 columns]

In [235]:
# Seleciona os atributos da base de dados
X = pd.DataFrame(database.iloc[:,0:20])
# Seleciona a classe da base de dados
y = pd.DataFrame(database.iloc[:,20])

Como o Random Forest não consegue processar dados que são categóricos, é necessário fazer a transformação deles em variáveis numéricas através do Label Encoder

In [236]:
# Índice das colunas que possuem informações categóricas
categorical_variables = [0, 2, 3, 5, 6, 8, 9, 11, 13, 14, 16, 18, 19]

# Laço de repetição que irá transformar as colunas categóricas em numéricas
for i in categorical_variables:
    X.iloc[:,i] = LabelEncoder().fit_transform(X.iloc[:,i])

In [237]:
# Divisão da base de dados entre treinamento e teste, sendo 70% para treinamento e 30% para teste
X_treinamento, X_teste, y_treinamento, y_teste = train_test_split(X,
                                                                  y,
                                                                  test_size = 0.3,
                                                                  random_state = 0)

In [238]:
# Criação e treinamento do modelo
random_forest = RandomForestClassifier(n_estimators = 100)
random_forest.fit(X_treinamento, y_treinamento)

C:\Users\francueudes\AppData\Local\Temp\ipykernel_11164\2639614448.py:3: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  random_forest.fit(X_treinamento, y_treinamento)


RandomForestClassifier()

In [239]:
# Previsões utilizando os registros de teste
predict = random_forest.predict(X_teste)
predict

array(['bad', 'good', 'good', 'good', 'good', 'good', 'good', 'good',
       'good', 'good', 'bad', 'good', 'good', 'good', 'bad', 'good',
       'good', 'good', 'bad', 'good', 'good', 'bad', 'good', 'good',
       'good', 'bad', 'good', 'good', 'good', 'bad', 'good', 'good',
       'good', 'good', 'good', 'bad', 'good', 'good', 'good', 'good',
       'good', 'bad', 'good', 'good', 'good', 'bad', 'good', 'bad',
       'good', 'good', 'good', 'bad', 'good', 'good', 'good', 'good',
       'bad', 'good', 'good', 'good', 'bad', 'good', 'good', 'good',
       'good', 'good', 'good', 'good', 'bad', 'good', 'good', 'good',
       'good', 'good', 'bad', 'good', 'good', 'good', 'good', 'good',
       'good', 'good', 'good', 'good', 'good', 'good', 'bad', 'good',
       'good', 'good', 'good', 'good', 'good', 'good', 'good', 'good',
       'good', 'good', 'good', 'good', 'bad', 'good', 'good', 'good',
       'good', 'good', 'bad', 'bad', 'good', 'good', 'good', 'bad',
       'good', 'good', 'goo

In [240]:
# Cria a matriz de confusão
confusion = confusion_matrix(y_teste, predict)
confusion

array([[ 39,  47],
       [ 16, 198]], dtype=int64)

In [241]:
# Obtêm a acurácia obtida pelo algoritmo de Naive Bayes
accuracy = accuracy_score(y_teste, predict)
accuracy

0.79

A acurácia obtida pelo algoritmo Random Forest foi de 79%